### PR TITLE
[export] Add custom op profiles and generate meta kernel

### PR DESCRIPTION
**Pull request by angelayi**
Example output:
```
1. Missing fake kernel.
    torch.ops.mylib.foo2.default is missing a fake kernel implementation.

    Please refer to https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit#heading=h.ahugy69p2jmz for more detailed instructions on how to implement a fake kernel.

    Here is an example of a fake kernel for your op, but it might not be correct for all use cases:
    ```
import torch.utils._pytree as pytree

@torch.library.register_fake("mylib::foo3")
def _(*args, **kwargs):
    ctx = torch.library.get_ctx()

    def match(tensor, dim, dtype, device, layout):
        # TODO: check strides
        return (
            tensor.dim() == dim
            and tensor.dtype == dtype
            and tensor.device == device
            and tensor.layout == layout
        )

    # Generate fake tensor for each output
    def get_fake_out(dim, sorted_strides, dtype, device, layout):
        fake_shape = [ctx.new_dynamic_size() for _ in range(dim)]

        fake_strides = [-1] * dim
        fake_stride = 1
        for idx in sorted_strides:
            fake_strides[idx] = fake_stride
            fake_stride = fake_stride * fake_shape[idx]

        return torch.empty_strided(
            size=fake_shape, stride=fake_strides, dtype=dtype, device=device, layout=layout
        )
    
    flat_args, args_spec = pytree.tree_flatten(args)
    
    if (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 3, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 3, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(3, [2, 1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    elif (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 2, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 2, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(2, [1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    else:
        raise NotImplementedError("Input type not seen before")
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140048
* #139384

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -2788,4 +2786,29 @@ def unsupported(reason: str) -> None:
...
+    def get_profile() -> Dict[str, Any]:
+        return {
+            "flat_args_profile": [maybe_get_tensor_meta(arg) for arg in flat_args],
+            "args_spec": pytree.treespec_dumps(args_spec),
+            "flat_out_profile": [maybe_get_tensor_meta(out) for out in real_flat_out],
+            "out_spec": pytree.treespec_dumps(spec),
+        }
+
+    dtrace_structured(
+        "generated_fake_kernel",
```
Comment (pianpwk) [inline torch/_subclasses/fake_tensor.py:2810]: nit: could we rename this case to "missing_fake_kernel" or something similar? https://github.com/pytorch/pytorch/pull/139766 will also generate fake kernels, but for the mismatched case

File: torch/export/_draft_export.py [hunk 2]
```diff
@@ -296,6 +384,34 @@ def draft_export(
...
+                        ]
+                    ),
+                    out_spec=op_profile["out_spec"],
+                )
+                seen_profiles.add(new_op_profile)
+
+            failures.append(
+                FailureReport(
+                    FailureType.MISSING_FAKE_KERNEL,
+                    {"op": op, "op_profiles": list(seen_profiles)},
```
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/export/_draft_export.py:411]: I was thinking we can use the op_profiles during retracing? Or in cases where we don't have real tensor tracing.

File: torch/export/_draft_export.py [hunk 3]
```diff
@@ -89,11 +91,63 @@ def __repr__(self) -> str:
...
+
+        # Loop through all the flat outputs from the profiling
+        flat_fake_out = [{call_get_fake_out_str}
+        ]
+
+        # Unflatten the outputs to be the same output structure as the eager model
+        return pytree.tree_unflatten(
+            flat_fake_out, pytree.treespec_loads(\'{op_profiles[0].out_spec}\')
+        )
+    ```
```
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/export/_draft_export.py:150]: Yeah for now I only used the first op profile.. it seemed a little hacky to generate the ifs, but thinking more about it, it shouldn't be too hard..

[the remark this answers is not in the document]
Comment (angelayi) [reply torch/export/_draft_export.py:199]: updated!

File: torch/export/_draft_export.py [hunk 4]
```diff
@@ -89,11 +98,105 @@ def __repr__(self) -> str:
...
+                strides = [(s, idx) for idx, s in enumerate(strides)]
+                strides.sort()
+                return [idx for _, idx in strides]
+
+            op_module, op_name, op_overload = op.split(".")
+
+            profile_casing = ""
+
+            for i, op_profile in enumerate(op_profiles):
+                check_arg_string = '            and match(flat_args[{idx}], {dim}, {dtype}, torch.device("{device}"), {layout})'
```
Comment (pianpwk) [inline torch/export/_draft_export.py:113]: nit: maybe add keywords? `size={dim}, dtype={dtype}, ...`, same for the bottom?

File: torch/export/_draft_export.py [hunk 5]
```diff
@@ -89,11 +98,105 @@ def __repr__(self) -> str:
...
+                    )
+                    for profile in op_profile.flat_out_profile
+                ]
+                call_get_fake_out_str = "\n".join(call_get_fake_out_list)
+
+                profile_casing += f"""
+        {"if" if i == 0 else "elif"} (
+            args_spec == pytree.treespec_loads(\'{op_profile.args_spec}\')
+{check_args_str}
+        ):
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/export/_draft_export.py:143]: Oh, is it the `_custom_ops_profile.py` that's still only in fbcode? But why is that not in OSS now? Sorry, I'm just confused.

**Review by pianpwk — APPROVED** [review]

File: torch/export/_draft_export.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/export/_draft_export.py:143]: the op profiles are created [further down in this PR](https://github.com/pytorch/pytorch/pull/140048/files#diff-86001d9642d74824f75bd856be741e67deb1baa8c4a07a43581f4ddd72e809f9R444-R460)

File: torch/export/_draft_export.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/export/_draft_export.py:465]: It's probably worth sitting down and ironing out the "ideal UX" for reuse of the custom op profile. Maybe the ideal UX is "we get the user to check in the meta to the codebase, so they do not need to serialize and pass the profile around".
